nemo/collections/nlp/modules/common/transformer/transformer.py

```diff
@@ -0,0 +1,177 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass
+
+from omegaconf.omegaconf import MISSING
+
+from nemo.collections.nlp.modules.common.decoder_module import DecoderModule
+from nemo.collections.nlp.modules.common.encoder_module import EncoderModule
+from nemo.collections.nlp.modules.common.transformer.transformer_decoders import TransformerDecoder
+from nemo.collections.nlp.modules.common.transformer.transformer_encoders import TransformerEncoder
+from nemo.collections.nlp.modules.common.transformer.transformer_modules import TransformerEmbedding
+
+
+@dataclass
+class TransformerConfig:
+    # must be configured by the user
+    hidden_size: int = MISSING
+    num_layers: int = MISSING
+    inner_size: int = MISSING
+    num_attention_heads: int = MISSING
+
+    # embedding
+    max_sequence_length: int = 512
+    num_token_types: int = 2
+    embedding_dropout: float = 0.0
+    learn_positional_encodings: bool = False
+
+    # transformer
+    ffn_dropout: float = 0.0
+    attn_score_dropout: float = 0.0
+    attn_layer_dropout: float = 0.0
+    hidden_act: str = 'relu'
+    pre_ln: bool = False
+
+
+@dataclass
+class TransformerEncoderConfig(TransformerConfig):
+    mask_future: bool = False
+
+
+class TransformerEncoderNM(EncoderModule):
```

Comment:
Do we need NM in the name?

Reply:
`TransformerEncoder` is already taken by the `nn.module`. I don't know of a better suffix. Any ideas?

Reply:
Prepend NMT? If they are the same names, we should be merging functionality not duplicating classes, but that's out of scope for now. 

We don't have any NM modules anymore (they were there in <v0.11)

Reply:
The neural module TransformerEncoderNM takes ids as input. The ids then go through TransformerEmbedding and TransformerEncoder so that the out of TransformerEncoderNM is hidden states. So the functionality is not duplicated, this let's us swap encoders with HF and Megatron for example. We can come up with a better name soon I guess.

Reply:
Sure sounds good, future PR can rename it. Leaving comment open for later review.